pkg/fwdmetrics/rate.go

```diff
@@ -0,0 +1,134 @@
+package fwdmetrics
+
+import (
+	"sync"
+	"time"
+)
+
+// DefaultMaxSamples is the default number of samples to keep in history
+const DefaultMaxSamples = 60 // 60 seconds of history
+
+// RateCalculator maintains a rolling window of samples for rate calculation
+type RateCalculator struct {
+	samples      []RateSample
+	maxSamples   int
+	currentIndex int
+	mu           sync.RWMutex
+}
+
+// NewRateCalculator creates a new rate calculator with the specified history size
+func NewRateCalculator(maxSamples int) *RateCalculator {
+	if maxSamples <= 0 {
+		maxSamples = DefaultMaxSamples
+	}
+	return &RateCalculator{
+		samples:    make([]RateSample, 0, maxSamples),
+		maxSamples: maxSamples,
+	}
+}
+
+// AddSample records a new sample for rate calculation
+func (rc *RateCalculator) AddSample(bytesIn, bytesOut uint64, timestamp time.Time) {
+	rc.mu.Lock()
+	defer rc.mu.Unlock()
+
+	sample := RateSample{
+		Timestamp: timestamp,
+		BytesIn:   bytesIn,
+		BytesOut:  bytesOut,
+	}
+
+	if len(rc.samples) < rc.maxSamples {
+		rc.samples = append(rc.samples, sample)
+	} else {
+		rc.samples[rc.currentIndex] = sample
+	}
+	rc.currentIndex = (rc.currentIndex + 1) % rc.maxSamples
+}
+
+// GetInstantRate returns bytes/sec based on last two samples
+func (rc *RateCalculator) GetInstantRate() (rateIn, rateOut float64) {
+	rc.mu.RLock()
+	defer rc.mu.RUnlock()
+
+	if len(rc.samples) < 2 {
+		return 0, 0
+	}
+
+	// Get current and previous samples (accounting for ring buffer)
+	currIdx := (rc.currentIndex - 1 + len(rc.samples)) % len(rc.samples)
+	prevIdx := (rc.currentIndex - 2 + len(rc.samples)) % len(rc.samples)
+
+	curr := rc.samples[currIdx]
+	prev := rc.samples[prevIdx]
+
+	duration := curr.Timestamp.Sub(prev.Timestamp).Seconds()
+	if duration <= 0 {
+		return 0, 0
+	}
+
+	rateIn = float64(curr.BytesIn-prev.BytesIn) / duration
+	rateOut = float64(curr.BytesOut-prev.BytesOut) / duration
```

Comment:
```suggestion
    // Handle counter wraparound or reset
    if curr.BytesIn < prev.BytesIn || curr.BytesOut < prev.BytesOut {
        return 0, 0
    }

	rateIn = float64(curr.BytesIn-prev.BytesIn) / duration
	rateOut = float64(curr.BytesOut-prev.BytesOut) / duration
```

If a pod is restarted, could it be possible that the new bytes counter is smaller than the old one? If so, this safety check will help.